src/sleep/sleep.c

```diff
@@ -263,41 +263,91 @@ static int execute(
-        if (r < 0)
-                return log_error_errno(errno, "Error setting hibernate timer: %m");
+                r = fd_wait_for_event(tfd, POLLIN, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error polling timerfd: %m");
+                /* Store fd_wait status */
+                woken_by_timer = FLAGS_SET(r, POLLIN);
+
+                r = read_battery_capacity_percentage();
+                if (r >= 0) {
+                        current_capacity = r;
+                        log_debug("Current battery charge percentage after wakeup: %d%%", current_capacity);
+                } else if (r == -ENOENT) {
+                        log_debug_errno(r, "Battery capacity percentage unavailable, cannot estimate discharge rate: %m");
+                        break;
+                        /* In case of no battery, system will be hibernated after 1st cycle of suspend */
+                } else
+                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
+
+                if (!woken_by_timer) {
+                        if (current_capacity < last_capacity) {
+                                /* We woke up before alarm time, estimate discharge rate using suspension duration. */
+                                after_timestamp = now(CLOCK_BOOTTIME);
+                                log_debug("Estimating discharge rate....");
+                                estimated_discharge_rate = ((last_capacity - current_capacity) * 60 * 60 * USEC_PER_SEC) / (after_timestamp - before_timestamp);
+                                /* The capacity difference is multiplied with 3600 to convert it to per hour */
+                                log_debug("Manual Wakeup. Battery discharge rate is %d%% per hour", estimated_discharge_rate);
+
+                        }
+                        return 0; /* return as manual wakeup done. This also will return in case battery was charged during suspension */
+                }
 
-        r = execute(sleep_config, SLEEP_SUSPEND, NULL);
-        if (r < 0)
-                return r;
+                if (current_capacity >= last_capacity)
+                        log_debug("Battery was not discharged during suspension");
+                else {
+                        log_debug("Attempting to estimate battery discharge rate after waking from %s hours sleep timer",
+                                  FORMAT_TIMESPAN(suspend_interval, USEC_PER_HOUR));
 
-        r = fd_wait_for_event(tfd, POLLIN, 0);
-        if (r < 0)
-                return log_error_errno(r, "Error polling timerfd: %m");
-        if (!FLAGS_SET(r, POLLIN)) /* We woke up before the alarm time, we are done. */
-                return 0;
+                        /* If woken up after alarm time, estimate discharge rate for suspend interval */
+                        estimated_discharge_rate = (last_capacity - current_capacity) / (suspend_interval * USEC_PER_HOUR);
```

Comment:
`USEC_PER_SEC` is in a wrong place...

Reply:
`(last_capacity - current_capacity) * USEC_PER_HOUR / suspend_interval`

Reply:
Please see the similar calculation you do in the above.

Reply:
oh my bad! so sorry. thanks. aah. i was contradicting myself by that mistake i made here. suspend_interval has to be in hour so yeah when it is divided, it goes in numerator. So sorry.